[camera_platform_interface] [camera] [camera_android] Add NV21 as an image stream format

acoutts opened the pull request:
### Note: this is a re-created PR from https://github.com/flutter/plugins/pull/6985
---

This PR adds NV21 as an image stream format for android devices. This is the format required for things like MLKit. Unfortunately a lot of people tend to implement the yuv->nv21 incorrectly in dart, which results in countless issues of users saying the image stream doesn't work, or that mlkit doesn't work.

By allowing the plugin to send NV21 to dart, this will fix all of those yuv conversion issues.

Highlights:

-   Added an abstraction around `ImageReader` called `ImageStreamReader` and moved the image stream logic into this class to clean up the `Camera` class. This allows us to test the image stream handler in isolation.
-   Added tests for `ImageStreamReader` to make sure it only calls the `removePlaneBufferPadding` function for planes that contain padding.
-   Added a new utility class called `ImageStreamReaderUtils` which contains the logic for trimming the padding.
-   Added tests for `ImageStreamReaderUtils` to make sure it only removes padding when a given buffer contains padding to be removed.
-   There are tests to confirm both that `removePlaneBufferPadding` is only called when padding is present, and to confirm that if `removePlaneBufferPadding` does happen to be called with a buffer containing no padding, it still returns a valid buffer.

*List which issues are fixed by this PR. You must list at least one issue.*

- Fixes https://github.com/flutter/flutter/issues/118350

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [ ] I signed the [CLA].
- [ ] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

camsim99 commented on packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/media/ImageStreamReader.java:
```diff
@@ -0,0 +1,223 @@
+
+    // We will convert the YUV data to NV21 which is a single-plane image
+    ByteBuffer bytes =
+        imageStreamReaderUtils.yuv420ThreePlanesToNV21(
```
Why not catch the illegal state exception here? Just wondering if this may not be the only cause.

acoutts replied:
I opted for doing it in the other way because I wanted to make sure we post an exception to dart and close the image after. If we catch it here then i'm not sure how we'd still return an error to dart because line 101 would presumedly return normally, which would then have the plugin try and post an image back to dart.

camsim99 commented on packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/media/ImageStreamReaderUtils.java:
```diff
@@ -0,0 +1,156 @@
+    byte[] out = new byte[imageSize + 2 * (imageSize / 4)];
+
+    if (areUVPlanesNV21(yuv420888planes, width, height)) {
+      //      Log.i("flutter", "planes are NV21");
```
Nit: remember to remove the logs!

acoutts replied:
Fixed in e0b57166b6c09a6a6886e002e734caf1db1f333e

camsim99 commented on packages/camera/camera_android/android/src/test/java/io/flutter/plugins/camera/media/ImageStreamReaderTest.java:
```diff
@@ -0,0 +1,164 @@
+    imageStreamReader.onImageAvailable(mockImage, mockCaptureProps, mockEventSink);
+
+    // Make sure we processed the frame with parsePlanesForNv21
+    verify(mockImageStreamReaderUtils).yuv420ThreePlanesToNV21(any(), anyInt(), anyInt());
```
Can you verify the parameters for this call too?

acoutts replied:
Fixed in c170fe0f7e2a22d1cfa6fb0fda9c77b085ff7c26

acoutts replied:
Fixed in c170fe0f7e2a22d1cfa6fb0fda9c77b085ff7c26

camsim99 reviewed (commented):
Overall, this looks great! Thanks for this and especially for wrapping `ImageReader` :) just left a few comments.

bparrishMines reviewed (changes requested):
This looks good to me, but the platform interface and android implementation are missing tests. You should be able to do something along the lines of: https://github.com/flutter/packages/blob/main/packages/camera/camera_platform_interface/test/method_channel/type_conversion_test.dart#L64

acoutts commented:
> This looks good to me, but the platform interface and android implementation are missing tests. You should be able to do something along the lines of: https://github.com/flutter/packages/blob/main/packages/camera/camera_platform_interface/test/method_channel/type_conversion_test.dart#L64

Tests added in cb8dc5e0aeef9b1df505f7310531ec8e5a3f0163

acoutts commented on packages/camera/camera/CHANGELOG.md:
```diff
@@ -1,3 +1,6 @@
+## 0.10.4
+* Adds NV21 as an image streaming option for Android.
```
Fixed in 093d9e7e53cd98cd8ab39c6d76a965ec8c6f234d

acoutts commented on packages/camera/camera_android/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 0.10.5
+
+* Adds support for NV21 as a new streaming format in Android which includes correct handling of image padding when present.
+
```
Fixed in 000e7834067dfe3156dde9c0bc68ec15e692e7bb

bparrishMines commented on packages/camera/camera/lib/src/camera_image.dart:
```diff
@@ -90,6 +90,9 @@ ImageFormatGroup _asImageFormatGroup(dynamic rawFormat) {
         return ImageFormatGroup.jpeg;
+      // android.graphics.ImageFormat.NV21
+      case 17:
+        return ImageFormatGroup.nv21;
```
This could also use a test. Probably in https://github.com/flutter/packages/blob/main/packages/camera/camera/test/camera_image_test.dart

acoutts replied:
Apologies - I thought I had that one, maybe I forgot to include it when I combined the PRs into one.

acoutts replied:
Fixed in 97787d5f5a619cd13a824e54ffccd4e55f346bb2

acoutts replied:
Merged main and removed tracked gradle files in 14f29861ce64a0ef7bfa3da3d7d8953b95d0f0e7

bparrishMines reviewed (approved):
LGTM with one more comment

It also looks like this was made before we updated the `.gitignore` from the repo merge. You will need to remove all the added gradle related files in `camera/example` and pull from main.

acoutts commented:
@bparrishMines i'm fixing one bug I've found today where sometimes imageStreamReader can be null when we try to subscribe a listener.

acoutts commented:
Ok that's pushed.

acoutts commented:
Merged main, fixed formatting issue and ran the android tests locally to confirm they are all passing. The one that failed before seems to be a JVM out of memory error. I made the test image smaller to fix that issue in the test.

bparrishMines commented:
@acoutts You can go ahead and create a separate PR that updates only the `camera_platform_interface` package.

acoutts commented:
Just opened https://github.com/flutter/packages/pull/3469

acoutts commented:
@bparrishMines saw it's published 🙌 
which would you prefer we do next - `camera` or `camera_android`?

acoutts commented:
Before opening that PR, i've just merged main into this branch and removed the dependency override for `camera_platform_interface` in `camera_android` now that 2.5.0 is published for that. If all looks good in `camera_android` I will open the PR for that.

xiaoshijiu1999 commented:
I wonder how I can modify pubspec.yaml to use the modified camera plugin.

farukprogrammer commented on packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/media/ImageStreamReader.java:
```diff
@@ -0,0 +1,223 @@
+  @VisibleForTesting
+  public static int computeStreamImageFormat(int dartImageFormat) {
+    if (dartImageFormat == ImageFormat.NV21) {
+      return ImageFormat.YUV_420_888;
```
Sorry I didn't get it, what's the point supporting NV21, but you change it to YUV? then you convert from YUV to NV21 again? we can do it in dart, and this is not solving the lag of imageStream. Can you enlight us?

acoutts replied:
The reasoning is discussed over in the original PR here: https://github.com/flutter/plugins/pull/6985

Camera2 can't stream NV21, only YUV so if you want it in NV21 you either convert it in dart or on the native side first. To make it easier for others to consume, the `camera_android` package handles the logic for this conversion and handles all of the edge cases such as when the frames contain buffer padding.

The lag you mention is probably just the nature of having to send things over the platform interface, YUV or NV21-converted. There will always be latency here compared to doing something purely on the native side.

Client-side image detection has latency to it even on the native side if you play with the google MLkit example app. The difference is when the latency is on the native side you get latency in the camera preview framerate, so your overlays always match up with the latency and it feels less noticeable. Latency between native and dart means the camera updates faster than the painted UI from Flutter and might feel worse because of that.

stuartmorgan commented:
@acoutts It looks like this should be ready to sync with `main` to make it landable?

acoutts commented:
Whooohoo! Cleaned it up and bumped the pubspec versions. One thing i'm confused on now, it looks like the test case for `bgra8888` in `camera_image_test` was overwritten / replaced with the nv21 one, so I fixed that as well.

caioCrmall commented:
Can this be added to stable? I really need it

caiofreitas commented:
Still facing this issue on these devices:
motorola edge 30 neo, poco x4 pro, poco x3 pro

caiofreitas commented:
> It's been merged for a while now. Make sure you're using nv21 on android too. It uses yuv by default.

It was exactly that.  My bad... 


CameraController(frontCamera, ResolutionPreset.ultraHigh,
        enableAudio: false, imageFormatGroup: ImageFormatGroup.nv21);
